src/compiler/utilities.ts

```diff
@@ -8374,4 +8374,37 @@ namespace ts {
         // '/// <reference no-default-lib="true"/>' directive.
         return options.skipLibCheck && sourceFile.isDeclarationFile || options.skipDefaultLibCheck && sourceFile.hasNoDefaultLib;
     }
+
+    export function isJsonEqual(a: unknown, b: unknown): boolean {
+        return a === b || typeof a === "object" && a !== null && typeof b === "object" && b !== null && equalOwnProperties(a as MapLike<unknown>, b as MapLike<unknown>);
+    }
+
+    export function getSourceFileAffectingCompilerOptions(compilerOptions: CompilerOptions): CompilerOptions {
+        const res: CompilerOptions = {};
+        for (const option of optionDeclarations) {
+            if (option.name in compilerOptions && isSourceFileAffectingCompilerOption(option)) {

```

Comment:
Like this! Can we instead put property on option declaration instead. Eg look at `affectsSemanticDiagnostics` in `CommandLineOptionBase`

Reply:
Also `noUnusedLabels` shouldnt need the new sourceFile. Instead it needs to set `affectsSemanticDiagnostics`

Reply:
It already had `affectsSemanticDiagnostics`, the problem is that the `--noUnusedLabels` error is attached to the `SourceFile#bindDiagnostics` directly during binding.  While `affectsSemanticDiagnostics` seems to only affect `semanticDiagnosticsPerFile`.